Diagnostics/FindFrontEndActivity.ps1

```diff
@@ -0,0 +1,198 @@
+.EXAMPLE
+    Get-ExchangeServer | .\FindFrontEndActivity.ps1 -SamAccountName "user1", "user2" -Quiet
+    Show only the server names where user1 or user2 are found in the MAPI HttpProxy logs within the last 15 minutes.
+.EXAMPLE
+    Get-ExchangeServer | .\FindFrontEndActivity.ps1 -SamAccountName "user1", "user2" -Protocol "ews", "mapi" -LatencyThreshold 100 -TimeSpan "00:30"
+    Show any EWS or MAPI HttpProxy activity that took more than 100 milliseconds for user1 or user2 within the last 30 minutes on all Exchange servers in the forest.
+#>
+[CmdletBinding(DefaultParameterSetName = 'Recent')]
+param (
+    [Parameter(Mandatory = $true, ValueFromPipeline = $true, ValueFromPipelineByPropertyName = $true)]
+    [Alias('Fqdn')]
+    [string[]]
+    $ServerName,
+
+    [Parameter(Mandatory = $true)]
+    [string[]]
+    $SamAccountName,
+
+    [Parameter()]
+    [int]
+    $LatencyThreshold = 1000,
+
+    [Parameter()]
+    [ValidateSet('Autodiscover', 'EAS', 'ECP', 'EWS', 'MAPI', 'OWA', 'PowerShell', 'RpcHttp')]
+    [string[]]
+    $Protocol = @('MAPI'),
+
+    [Parameter()]
+    [switch]
+    $IncludeNonExecutes,
+
+    [Parameter()]
+    [switch]
+    $Quiet,
+
+    [Parameter(ParameterSetName = 'Recent')]
+    [TimeSpan]
+    $TimeSpan = (New-TimeSpan -Minutes 15),
+
+    [Parameter(ParameterSetName = 'Range', Mandatory = $true)]
+    [DateTime]
+    $StartTime,
+
+    [Parameter(ParameterSetName = 'Range', Mandatory = $true)]
+    [DateTime]
+    $EndTime
+)
+
+begin {
+    $serverNames = New-Object System.Collections.ArrayList
+    $samAccountRegexForFileMatch = [string]::Join('|', ($SamAccountName | ForEach-Object { "\\$_," }))
+    $samAccountRegexForFieldMatch = [string]::Join('|', ($SamAccountName | ForEach-Object { "\\$_$" }))
+}
+
+process {
+    foreach ($name in $ServerName) {
+        [void]$serverNames.Add($name)
+    }
+}
+
+end {
+    $timeThreshold = $null
+    if ($PSCmdlet.ParameterSetName -eq 'Recent') {
+        $timeThreshold = (Get-Date).Add(-$TimeSpan)
+    }
+
+    $jobs = $serverNames | ForEach-Object {
+        Invoke-Command -AsJob -ComputerName $_ -ScriptBlock {
+            $exchangeInstallPath = (Get-ItemProperty -Path Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\ExchangeServer\v15\Setup -ErrorAction SilentlyContinue).MsiInstallPath
+            $httpProxyLogPath = Join-Path $exchangeInstallPath "Logging\HttpProxy"
+            $results = New-Object System.Collections.ArrayList
+            foreach ($thisProtocol in $using:Protocol) {
+                $logPathForThisProtocol = Join-Path $httpProxyLogPath $thisProtocol
+                if (-not (Test-Path $logPathForThisProtocol)) {
+                    continue
+                }
+
+                $files = Get-ChildItem "$logPathForThisProtocol\*.log" | Where-Object {
+                    if ($null -ne $using:timeThreshold) {
+                        $_.LastWriteTime -gt $using:timeThreshold
```

Comment:
Not sure how common it is, but possible since this script doesn't require to be run on the server itself you can have different time zones from the execution computer and the Exchange Servers or possibly have different time zones within all the servers themselves. Best to use UTC.

Reply:
That shouldn't be necessary. DateTime contains an offset. I just tested this in my lab, and it pulls the correct file:

![image](https://user-images.githubusercontent.com/4518572/218501976-d63375b3-8ae2-4f57-96db-eed87bcbd7b9.png)

Reply:
Perfect, didn't know that.